File: 2020/4.php
<?php

require('libs/core.php');

function get_records($data) {
    $single = [];
    $records = [];
    foreach ($data as $i => $line) {
        if (empty(trim($line))) {
            $records[] = implode(" ", $single);
            $single = [];
        } else {
            $single[] = $line;
        }
    }
    $records[] = implode(" ", $single);
    return $records;
}

function validate_record($line) {
    if (preg_match_all('/(\w{3}):(\S+)/', $line, $matches)) {
        $fields = array_unique($matches[1]);
        if ((count($fields) >= 7) && ((count($fields) == 8) || (!in_array("cid", $fields)))) {
            return [true, $matches];
        }
    }
}

function validate_1($data) {
    $total = 0;
    foreach ($data as $i => $line) {
        if (validate_record($line)[0]) {
            $total++;
        }
    }
    print "\n{$total} valid\n";
}

function validate_2($data) {
    $total = 0;
    $hgt = ['cm' => [150, 193], 'in' => [59, 76]];
    foreach ($data as $i => $line) {
        list($valid, $matches) = validate_record($line);
        if ($valid) {
            $fields = array_combine($matches[1], $matches[2]);
            $valid = true;
            foreach ($fields as $key => $value) {
                switch ($key) {
                    case 'byr':
                        if (!preg_match('/^\d{4}$/', $value) || $value < 1920 || $value > 2002) {
                            $valid = false;
                        }
                        break;
                    case 'iyr':
                        if (!preg_match('/^\d{4}$/', $value) || $value < 2010 || $value > 2020) {
                            $valid = false;
                        }
                        break;
                    case 'eyr':
                        if (!preg_match('/^\d{4}$/', $value) || $value < 2020 || $value > 2030) {
                            $valid = false;
                        }
                        break;
                    case 'hgt':
                        if (!preg_match('/^(\d+)(cm|in)$/', $value, $matches)
                            || $matches[1] < $hgt[$matches[2]][0]
                            || $matches[1] > $hgt[$matches[2]][1]) {
                            $valid = false;
                        }
                        break;
                    case 'hcl':
                        if (!preg_match('/^#([\da-f]{6})$/', $value)) {
                            $valid = false;
                        }
                        break;
                    case 'ecl':
                        if (!in_array($value, ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'])) {
                            $valid = false;
                        }
                        break;
                    case 'pid':
                        if (!preg_match('/^\d{9}$/', $value)) {
                            $valid = false;
                        }
                        break;
                }
            }
            if ($valid) {
                $total++;
            }
        }
    }
    print "\n{$total} valid\n";
}

$data = load_data("4.txt");

$data = get_records($data);

validate_1($data);

validate_2($data);

File: 2020/python/4.py
import core
import re

def get_records(data):
	single = []
	records = []
	for line in data:
		if len(line.strip()) == 0:
			records.append(" ".join(single))
			single = []
		else:
			single.append(line)
	records.append(" ".join(single))
	return records

def validate_record(line):
	matches = re.findall('(\w{3}):(\S+)', line)
	if (matches == None) or len(matches) < 7:
		return [False]
	fields = []
	for match in matches:
		fields.append(match[0])
	if (len(fields) == 8) or not ("cid" in fields):
		return [True, matches]
	return [False]

def validate_1(data):
	total = 0
	for line in data:
		response = validate_record(line)
		if response[0]:
			total += 1
	print("%d valid" % total)

def validate_2(data):
	total = 0
	hgt = {'cm' : [150, 193], 'in' : [59, 76]}
	for line in data:
		response = validate_record(line)
		if response[0]:
			valid = True
			for match in response[1]:
				if match[0] == "byr":
					if (re.match('^\d{4}$', match[1]) == None) or (int(match[1]) < 1920) or (int(match[1]) > 2002):
						valid = False
				elif match[0] == "iyr":
					if (re.match('^\d{4}$', match[1]) == None) or (int(match[1]) < 2010) or (int(match[1]) > 2020):
						valid = False
				elif match[0] == "eyr":
					if (re.match('^\d{4}$', match[1]) == None) or (int(match[1]) < 2020) or (int(match[1]) > 2030):
						valid = False
				elif match[0] == "hgt":
					m = re.match('^(\d+)(cm|in)$', match[1])
					if (m == None) or (int(m.group(1)) < hgt[m.group(2)][0]) or (int(m.group(1)) > hgt[m.group(2)][1]):
						valid = False
				elif match[0] == "hcl":
					m = re.match('^#([\da-f]{6})$', match[1])
					if m == None:
						valid = False
				elif match[0] == "ecl":
					if not match[1] in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']:
						valid = False
				elif match[0] == "pid":
					m = re.match('^\d{9}$', match[1])
					if m == None:
						valid = False
			if valid:
				total += 1
	print("%d valid" % total)

data = core.load_data("4.txt")

data = get_records(data);

validate_1(data);

validate_2(data);
